README.md
<h6 align=center>
<img src="https://github.com/opsdroid/style-guidelines/raw/master/logos/logo-wide-light.png" alt="Opsdroid Logo"/>
</h6>

<h4 align=center>A browser based web app for chatting with <a href="https://github.com/opsdroid/opsdroid">opsdroid</a>.</h4>

<a href="https://github.com/opsdroid/opsdroid-web/actions"><img src="https://github.com/opsdroid/opsdroid-web/actions/workflows/node.js.yml/badge.svg" alt="Github CI Status"></img></a>
![GitHub](https://img.shields.io/github/license/opsdroid/opsdroid-web)
![GitHub last commit](https://img.shields.io/github/last-commit/opsdroid/opsdroid-web)
<a href="https://app.element.io/#/room/#opsdroid-general:matrix.org"><img src="https://img.shields.io/matrix/opsdroid-general:matrix.org.svg?logo=matrix" alt="Matrix Chat" /></a>

---

<p align="center">
  <a href="https://docs.opsdroid.dev/en/stable/quickstart.html">Quick Start</a> •
  <a href="https://docs.opsdroid.dev">Documentation</a> •
  <a href="https://playground.opsdroid.dev">Playground</a> •
  <a href="https://medium.com/opsdroid">Blog</a> •
  <a href="https://app.element.io/#/room/#opsdroid-general:matrix.org">Community</a>
</p>

---

<img width="499" alt="Screenshot 2022-06-19 at 20 15 14" src="https://user-images.githubusercontent.com/3131401/174496947-dcd648e1-14ea-4d44-9850-f058d85e8054.png">

## Contributors

This project exists thanks to all the people who contribute. Read the [contributing docs](./CONTRIBUTING.md) to know how to contribute to the project.
<a href="https://github.com/opsdroid/opsdroid-web/graphs/contributors"><img src="https://opencollective.com/opsdroid/contributors.svg?width=890" /></a>


CHANGELOG.md
# Opsdroid Web Release Notes

These release notes are related to updates to the `opsdroid-web` package.

## v0.2.3

- Documentation changes
- Fixing issues for npm release

## v0.2.2

- Dependencies updates
- Dropped support for node versions 12 and 14
- Allow users using Chrome to send commands using voice using the speech recognition feature.

## v0.2.1

- Fix the theme defaulting to the dark mode - the app will now look into your prefered theme
- You can now submit a profile picture to be used when you send messages to opsdroid
- Error messages can now be closed

## v0.2.0

- Rewrite the whole package using React/Typescript
  - Add settings so you can control different parts of the app
  - The app was redesigned to look more modern
  - Add the possibility to change the theme (dark/light) and accent colours (blue/green)
  - Add the possibility to set a token to be used when authenticating to opsdroid websocket
  - You can now choose your username to be sent to opsdroid
- Add messages sent history so you can use arrow up/down to cycle through sent messages
  -Expose connection errors so you know when and why the app failed to connect to Opsdroid

## v0.1.0

- Split up web app from `opsdroid-desktop`


webpack.config.js
const HtmlWebPackPlugin = require("html-webpack-plugin");

module.exports = {
    module: {
      rules: [
        {
          test: /\.(js|jsx)$/,
          exclude: /node_modules/,
          use: {
            loader: "babel-loader"
          }
        },
        {
          test: /\.html$/,
          use: [
            {
              loader: "html-loader"
            }
          ]
        },
        {
            test:/\.(s*)css$/,
            use: [
                "style-loader", // creates style nodes from JS strings
                "css-loader", // translates CSS into CommonJS
                "sass-loader" // compiles Sass to CSS, using Node Sass by default
            ]
        },
        {
          test: /\.(jpe?g|png|gif|svg)$/i,
          loaders: [
            'file-loader',
            'image-webpack-loader'
          ]
        }
      ]
    },
    node: {
        fs: "empty",
        tls: "empty",
        net: "empty"
    },
    devtool: 'source-map',
    plugins: [
        new HtmlWebPackPlugin({
            template: "./src/index.html",
            filename: "./index.html"
        })
    ]
  };


src/store/index.ts
import { Store } from "pullstate";
import * as settings from "browser-cookies";
import { MessageType } from "../types";
import { WebsocketConnector } from "../websockets";

export type DisconnectedType = { type: "disconnected" };
export type ConnectingType = { type: "connecting" };
export type ConnectedType = { type: "connected" };
export type ConnectionErrorType = {
  type: "error";
  error: string;
  data?: object | string;
};

export type PartialConnectionState = {
  loadState: ConnectionState;
  client?: WebsocketConnector;
  connected: boolean;
};

export type ClientSettings = {
  host: string;
  port: string;
  sslEnabled: boolean;
  showSettings: boolean;
  token?: string;
};

export type ConnectionState =
  | DisconnectedType
  | ConnectingType
  | ConnectedType
  | ConnectionErrorType;

export type Connection = {
  loadState: ConnectionState;
  cooldown: number;
  timeout: number;
  client?: WebsocketConnector;
  connected: boolean;
};

export type Appearance = {
  darkTheme: boolean;
  accentColor: "blue" | "green" | string;
};

export type UserSettings = {
  username: string;
  avatar?: string;
};

export type AppState = {
  clientSettings: ClientSettings;
  connection: Connection;
  conversation: Array<MessageType>;
  userSettings: UserSettings;
  appearance: Appearance;
};

let accentColor;
const storedAccentColor = settings.get("accent-color");

if (storedAccentColor && storedAccentColor === "green") {
  accentColor = "green";
} else {
  accentColor = "blue";
}

export const UIStore = new Store<AppState>({
  clientSettings: {
    host: settings.get("host") || "localhost",
    port: settings.get("port") || "8080",
    sslEnabled: settings.get("sslEnabled") === "true" || false,
    showSettings: false,
    token: settings.get("token") || undefined,
  },
  connection: {
    loadState: { type: "disconnected" },
    cooldown: 1,
    timeout: 1000,
    client: undefined,
    connected: false,
  },
  conversation: [],
  appearance: {
    darkTheme: settings.get("dark-theme") == "true" ? true : false,
    accentColor: accentColor,
  },
  userSettings: {
    username: settings.get("username") || "user",
    avatar: localStorage.getItem("avatar") || undefined,
  },
});

// Reactions for state change

UIStore.createReaction(
  (s) => s.clientSettings,
  (original) => {
    // Note: Here we are just adding any changes to the cookies, it's likely
    // that we don't really need to handle conditionals here, since if only
    // one of the settings is changed, the other one will be updated with the
    // original value.
    settings.set("host", original.host);
    settings.set("port", String(original.port));
    settings.set("sslEnabled", String(original.sslEnabled));
    if (original.token) {
      settings.set("token", String(original.token));
    }
  }
);

UIStore.createReaction(
  (s) => s.conversation,
  (watched, draft, original, lastWatched) => {
    if (watched.length > lastWatched.length) {
      const lastMessage = watched[watched.length - 1];
      if (lastMessage.user !== "opsdroid" && original.connection.client) {
        original.connection.client.send(lastMessage);
      }
    }
  }
);

UIStore.createReaction(
  (s) => s.userSettings,
  (original) => {
    settings.set("username", original.username);
    if (original.avatar) {
      localStorage.setItem("avatar", original.avatar);
    }
  }
);

UIStore.createReaction(
  (s) => s.appearance,
  (original) => {
    settings.set("dark-theme", String(original.darkTheme));
    settings.set("accent-color", original.accentColor);
  }
);


src/js/components/UpdateMessage.js
'use strict';

// import electron from 'electron';
import package_info from '../../../package.json';


//////
// Imports
import React from 'react';


export default class UpdateMessage extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      latestVersion: false,
    };

    this.handleUpdateCheck = this.handleUpdateCheck.bind(this);
    this.handleClick = this.handleClick.bind(this);
  }

  render() {
    if (this.state.latestVersion) {
      return (
        <div className="updates-available" onClick={this.handleClick}>
          Update available. Download {this.state.latestVersion["name"]} now.
        </div>
      );
    }
    return false;
  }

  componentDidMount() {
    // this.checkForUpdates(this.handleUpdateCheck);
  }

  handleUpdateCheck(updatesAvailable) {
    this.setState({latestVersion: updatesAvailable})
  }

  handleClick(event) {
    event.preventDefault();
    // electron.shell.openExternal(this.state.latestVersion["html_url"]);
  }

  checkForUpdates(callback) {
    const RELEASES_URL = 'https://api.github.com/repos/opsdroid/opsdroid-desktop/releases';
    var myInit = { method: 'GET',
                   headers: new Headers(),
                   mode: 'cors',
                   cache: 'default' };
    var myRequest = new Request(RELEASES_URL, myInit);
    fetch(myRequest).then(function(response) {
      var contentType = response.headers.get("content-type");
      if(contentType && contentType.indexOf("application/json") !== -1) {
        return response.json().then(function(json) {
          let latest_version = json[0];
          if (latest_version["name"] != 'v' + package_info["version"]) {
            console.log("An update is available.");
            callback(latest_version);
          } else {
            console.log("No updates available.");
            callback(false);
          }
        });
      } else {
        console.log("Got a non json response from GitHub.");
      }
    }).catch(function(){
      console.log("Oops, could not check for latest release! No internet.");
    });
  }
}


src/index.js
'use strict';


//////
// Imports
import React from 'react';
import ReactDOM from 'react-dom';
import request from 'request';
import * as settings from 'browser-cookies';
import WebSocket from 'websocket';

import {formatPostUrl, formatSocketUrl, checkForUrl} from './js/utils';

import Conversation from './js/components/Conversation';
import Prompt from './js/components/Prompt';
import ConnectionSettings from './js/components/ConnectionSettings';
import UpdateMessage from './js/components/UpdateMessage';

import './css/index.scss'


//////
// Constants
const DEFAULT_HOST = "localhost";
const DEFAULT_PORT = "8080";
const DEFAULT_SSL_PORT = "8443";
const DEFAULT_SSL_ENABLED = false;
const DEFAULT_COOKIE_CONFIG = {expires: 365};


//////
// Main Class
class ChatClient extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      conversation: [],
      connected: false,
      showConnectionSettings: false,
      host: settings.get("host") || DEFAULT_HOST,
      port: settings.get("port") || DEFAULT_PORT,
      sslEnabled: settings.get("sslEnabled") === 'true' || DEFAULT_SSL_ENABLED,
    };

    this.active_connection = undefined;
    this.client = undefined;
    this.connected = false;
    this.connectionCooldown = 0;
    this.connectionTimeout = undefined;

    //////
    // Event listeners
    this.toggleConnectionSettings = this.toggleConnectionSettings.bind(this);
    this.sendUserMessage = this.sendUserMessage.bind(this);
    this.updateHost = this.updateHost.bind(this);
    this.updatePort = this.updatePort.bind(this);
    this.toggleSSL = this.toggleSSL.bind(this);
    this.connectToWebsocket = this.connectToWebsocket.bind(this);
    this.reconnectToWebSocketImmediately = this.reconnectToWebSocketImmediately.bind(this);

    this.handleSocketConnect = this.handleSocketConnect.bind(this);
    this.handleSocketFailedToConnect = this.handleSocketFailedToConnect.bind(this);
    this.handleSocketError = this.handleSocketError.bind(this);
    this.handleSocketClose = this.handleSocketClose.bind(this);
    this.handleSocketMessage = this.handleSocketMessage.bind(this);
  }

  render() {
    return (
      <div>
        <UpdateMessage />
        <Conversation
          items={this.state.conversation} />
        <Prompt
          connected={this.state.connected}
          toggleConnectionSettings={this.toggleConnectionSettings}
          sendUserMessage={this.sendUserMessage} />
        <ConnectionSettings
          host={this.state.host}
          port={this.state.port}
          sslEnabled={this.state.sslEnabled}
          defaultHost={DEFAULT_HOST}
          defaultPort={DEFAULT_PORT}
          visible={this.state.showConnectionSettings}
          updateHost={this.updateHost}
          updatePort={this.updatePort}
          toggleSSL={this.toggleSSL}
          reconnectToWebSocketImmediately={this.reconnectToWebSocketImmediately} />
      </div>
    );
  }

  componentDidMount() {
    this.connectToWebsocket();
  }

  handleSocketConnect() {
    console.log('WebSocket Client Connected');
    this.active_connection = this.client;
    this.resetCooldown();
    this.hideConnectionSettings();
    this.updateStatusIndicator(true);
    this.addMessage('connected', 'info');

    this.client.onerror = this.handleSocketError;
    this.client.onclose = this.handleSocketClose;
    this.client.onmessage = this.handleSocketMessage;
  }

  handleSocketFailedToConnect(error) {
    console.log('Connect Error: ' + error.toString());
    this.reconnectToWebSocket();
  }

  handleSocketError(error) {
    console.log("Connection Error: " + error.toString());
    this.updateStatusIndicator(false);
    this.addMessage('connection error', 'info');
  }

  handleSocketClose() {
    console.log('echo-protocol Connection Closed');
    this.updateStatusIndicator(false);
    this.reconnectToWebSocket();
    this.addMessage('disconnected', 'info');
  }

  handleSocketMessage(message) {
    if (typeof message.data === 'string') {
        this.addMessage(message.data, "bot");
    }
  }

  addMessage(message, sender) {
    // Add new message to the conversation
    let newMessage = {
      "text": message,
      "user": sender,
      "time": new Date()
    }
    this.setState((prevState, props) => ({
      conversation: prevState.conversation.concat([newMessage])
    }));
    this.expandMessage(newMessage);
  }

  expandMessage(message) {
    let url = checkForUrl(message.text);
    if (url) {
      let req = {
        method: 'GET', headers: new Headers(),
        mode: 'cors', cache: 'default'};
      fetch(new Request(url, req)).then((response) => {
        return response.blob();
      }).then((blob) => {
        if (blob.type == "image/gif" ||
            blob.type == "image/png" ||
            blob.type == "image/jpg" ||
            blob.type == "image/jpeg") {
          this.setState((prevState, props) => {
            let index = prevState.conversation.indexOf(message);
            prevState.conversation[index].image = url;
            return {"conversation": prevState.conversation};
          });
        }
      })
    }
  }

  updateStatusIndicator(status) {
    this.setState({connected: status});
  }

  sendUserMessage(user_message) {
    if (this.client.readyState === this.client.OPEN) {
      if (user_message != "") {
        this.addMessage(user_message, "user");
        this.sendMessageToSocket(user_message);
      }
    }
  }

  connectToWebsocket() {
    request.post(formatPostUrl(this.state.host, this.state.port, this.state.sslEnabled), (error, response, body) => {
      if (error) {
        console.log(error);
        this.addMessage('failed to connect', 'info');
        this.reconnectToWebSocket();
      } else {
        console.log(body)
        var socket = JSON.parse(body)["socket"];
        this.client = new WebSocket.w3cwebsocket(formatSocketUrl(this.state.host, this.state.port, this.state.sslEnabled, socket));
        this.client.onopen = this.handleSocketConnect
      }
    })
  }

  reconnectToWebSocket() {
    if (this.active_connection && this.active_connection.connected) {
      this.active_connection.close();
    }
    if (this.connectionTimeout) {
      clearTimeout(this.connectionTimeout);
    }
    console.log(`Reconnecting in ${this.connectionCooldown} seconds.`);
    this.connectionTimeout = setTimeout(this.connectToWebsocket, this.connectionCooldown * 1000);
    this.backoffCooldown();
  }

  reconnectToWebSocketImmediately() {
    this.addMessage('reconnecting', 'info');
    this.resetCooldown();
    this.reconnectToWebSocket();
  }

  resetCooldown() {
    this.connectionCooldown = 0;
  }

  backoffCooldown() {
      if (this.connectionCooldown <1 ) {
        this.connectionCooldown = 1;
      } else if (this.connectionCooldown < 5) {
        this.connectionCooldown = this.connectionCooldown * 2;
      }
  }

  sendMessageToSocket(message) {
    if (this.client.readyState === this.client.OPEN) {
        console.log("Sending " + message);
        this.client.send(message);
    } else {
        console.log("Unable to send message");
    }
  }

  toggleConnectionSettings() {
    this.setState((prevState, props) => ({
      showConnectionSettings: ! prevState.showConnectionSettings
    }));
  }

  hideConnectionSettings() {
    this.setState({
      showConnectionSettings: false
    });
  }

  updateHost(event) {
    if (event.target.value == '') {
      settings.set("host", DEFAULT_HOST, DEFAULT_COOKIE_CONFIG);
      this.setState({host: DEFAULT_HOST});
    } else {
      settings.set("host", event.target.value, DEFAULT_COOKIE_CONFIG);
      this.setState({host: event.target.value});
    }
    this.reconnectToWebSocketImmediately();
  }

  updatePort(event) {
    if (event.target.value == '') {
      settings.set("port", DEFAULT_PORT, DEFAULT_COOKIE_CONFIG);
      this.setState({port: DEFAULT_PORT});
    } else {
      settings.set("port", event.target.value, DEFAULT_COOKIE_CONFIG);
      this.setState({port: event.target.value});
    }
    this.reconnectToWebSocketImmediately();
  }

  toggleSSL(event) {
    this.setState((prevState, props) => {
      let sslEnabled = !prevState.sslEnabled
      settings.set("sslEnabled", sslEnabled ? "true": "false", DEFAULT_COOKIE_CONFIG);
      return {sslEnabled: !prevState.sslEnabled};
    });

    this.reconnectToWebSocketImmediately();
  }
}

//////
// Start
ReactDOM.render(<ChatClient />,
  document.getElementById("wrapper")
);
